### PR TITLE
Chronos : Add tune for AutoformerForecaster

### PR DESCRIPTION
**Pull request by rnwang04**
## Description
This PR want to add tune function  for AutoformerForecaster. By the way, enable seed search in AutoformerForecaster and fix epoch reset bug in AutoformerForecaster and BaseForecaster(https://github.com/intel-analytics/BigDL/issues/5062).

## Motivation and Context
AutoformerForecaster's tune func is different from BaseForecaster.

## Related Link or Issue
https://github.com/intel-analytics/BigDL/issues/5062

## API Usage or Code Design

```
forecaster = AutoformerForecaster(past_seq_len=24,
                                          future_seq_len=5,
                                          input_feature_num=2,
                                          output_feature_num=2,
                                          label_len=12,
                                          freq='s',
                                          loss="mse",
                                          metrics=['mae', 'mse', 'mape'],
                                          lr=space.Real(0.001, 0.01, log=True))
forecaster.tune(train_data, validation_data=val_data, n_trials=2)
forecaster.fit(train_data, epochs=3, batch_size=32)
evaluate = forecaster.evaluate(val_data)
```

## How was this PR tested?
- [x] unit tests
- [x] jenkins passed (please provide link):
http://10.112.231.51:18889/job/BigDL-Chronos-PR-Validation/559/

### CODE REVIEW

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py [hunk 1]
```diff
@@ -131,9 +135,14 @@ def __init__(self,
...
 
+        self.metrics = metrics
+
         self.distributed = distributed
-        self.seed = seed
         self.checkpoint_callback = True
+        # seed setting
+        if not isinstance(seed, Space):
+            from pytorch_lightning import seed_everything
+            seed_everything(seed=seed, workers=True)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py:145]: just a question here: do we need to set `workers=True` for other forecasters?(TCN, S2S, LSTM...)

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py [hunk 2]
```diff
@@ -142,13 +151,149 @@ def __init__(self,
...
+        self.use_hpo = True
 
-        # seed setting
-        from pytorch_lightning import seed_everything
-        seed_everything(seed=self.seed)
+        has_space = self._config_has_search_space(
+            config={**self.model_config, **self.optim_config,
+                    **self.loss_config, **self.data_config})
+
+        if not self.use_hpo and has_space:
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py:160]: it seems that this condition will not be passed under any circumstances since `self.use_hpo` is always True.

**Review by TheaperDeng — COMMENTED** [review]
more comments to be added

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py [hunk 3]
```diff
@@ -142,13 +151,149 @@ def __init__(self,
...
+             direction,
+             n_trials=2,
+             n_parallels=1,
+             epochs=1,
+             batch_size=32,
+             **kwargs):
+        """
+        Search the hyper parameter.
+
+        :param data: train data, as numpy ndarray tuple (x, y)
```
[the remark this answers is not in the document]
Comment (TheaperDeng) [reply python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py:222]: btw, we may support dataloader input here, since numpy tuple input is really mem consuming in some circumstances.

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py (hunk 1 above)
Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py:145]: Set workers=True in [seed_everyhing](https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.utilities.seed.html?highlight=seed_everything#pytorch_lightning.utilities.seed.seed_everything)  will properly configure all dataloaders passed to the Trainer with a worker_init_fn.
I think this might be a more fixed and reproducible way and can set to other forecasters. But more exp of this parameter can be tested.

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py (hunk 2 above)
Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py:160]: done. remove this check.

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py [hunk 4]
```diff
@@ -142,13 +151,149 @@ def __init__(self,
...
+    def _config_has_search_space(config):
+        """Check if there's any search space in configuration."""
+        for _, v in config.items():
+            if isinstance(v, Space):
+                return True
+            if isinstance(v, list):
+                for item in v:
+                    if isinstance(item, Space):
+                        return True
+        return False
```
[the remark this answers is not in the document]
Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py:183]: done

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py]: > I don't think data is a numpy tuple (x, y) here.

done. change to (x, y, x_enc, y_enc)

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py [hunk 5]
```diff
@@ -142,13 +151,149 @@ def __init__(self,
...
+        """
+        Search the hyper parameter.
+
+        :param data: train data, as numpy ndarray tuple (x, y)
+        :param validation_data: validation data, as numpy ndarray tuple (x,y)
+        :param target_metric: the target metric to optimize,
+               a string or an instance of torchmetrics.metric.Metric
+        :param direction: in which direction to optimize the target metric,
+               "maximize" - larger the better
+               "minimize" - smaller the better
```
[the remark this answers is not in the document]
Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py:213]: done

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py (hunk 1 above)
Comment (TheaperDeng) [reply python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py:152]: sure, we will have some quick check on this param in future.

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py [hunk 6]
```diff
@@ -148,13 +158,134 @@ def __init__(self,
...
-        # seed setting
-        from pytorch_lightning import seed_everything
-        seed_everything(seed=self.seed)
+        has_space = _config_has_search_space(
+            config={**self.model_config, **self.optim_config,
+                    **self.loss_config, **self.data_config})
+
+        if not has_space:
+            if self.use_hpo:
+                warnings.warn("HPO is enabled but no spaces is specified, so disable HPO.")
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py:169]: Small comment(you may ignore this to next PR): This logic will cause those users who just want to use fixed parameters in Autoformer always see this warning.

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py [hunk 7]
```diff
@@ -185,10 +316,13 @@ def fit(self, data, epochs=1, batch_size=32):
...
 
+        from bigdl.chronos.pytorch import TSTrainer as Trainer
         # Trainer init and fitting
-        self.trainer = Trainer(logger=False, max_epochs=epochs,
-                               checkpoint_callback=self.checkpoint_callback, num_processes=1,
-                               use_ipex=self.use_ipex, distributed_backend="spawn")
+        if not self.use_hpo:
+            self.trainer = Trainer(logger=False, max_epochs=epochs,
+                                   checkpoint_callback=self.checkpoint_callback, num_processes=1,
+                                   use_ipex=self.use_ipex, distributed_backend="spawn")
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py:324]: what will happen if users call `fit` but use same configspace during the initialization? If it's not allowed, we may give clear error message here.

File: python/chronos/src/bigdl/chronos/model/autoformer/Autoformer.py [hunk 8]
```diff
@@ -151,19 +149,15 @@ def forward(self, x_enc, x_mark_enc, x_dec, x_mark_dec,
...
         outputs = outputs[:, -self.pred_len:, :]
         batch_y = batch_y[:, -self.pred_len:, :]
         return self.loss(outputs, batch_y)
 
     def validation_step(self, batch, batch_idx):
         batch_x, batch_y, batch_x_mark, batch_y_mark = map(lambda x: x.float(), batch)
-        dec_inp = torch.zeros_like(batch_y[:, -self.pred_len:, :]).float()
-        dec_inp = torch.cat([batch_y[:, :self.label_len, :], dec_inp], dim=1).float()
-        outputs = self(batch_x.float(), batch_x_mark.float(), dec_inp,
+        outputs = self(batch_x.float(), batch_x_mark.float(), batch_y,
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/model/autoformer/Autoformer.py:160]: do we still need these `.float()`?

**Review by TheaperDeng — COMMENTED** [review]
others LGTM

File: python/chronos/src/bigdl/chronos/model/autoformer/Autoformer.py (hunk 8 above)
Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/model/autoformer/Autoformer.py:160]: don't need anymore. I will remove these redundant code.

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py (hunk 6 above)
Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py:169]: may be solved in next PR : )

File: python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py (hunk 7 above)
Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/forecaster/autoformer_forecaster.py:324]: done. Add a check and a ut for this case.
